doc/design/python-api-1.md

```diff
@@ -0,0 +1,187 @@
+   * `Asset`:
+        * `path: str`
+        * `sha256: str` — Should this only be on `RemoteAsset`?
+        * `size: int`
+        * `created: datetime` — Should this only be on `RemoteAsset`?
+        * `modified: datetime` — Should this only be on `RemoteAsset`?
+        * `get_metadata() -> BareAssetMeta`
+
+* `dandi.dandiapi`:
+    * Upon creating a new `RESTFullAPIClient` instance, a new session object should automatically be created and used for all of the client's requests; the `session()` method should thus be eliminated
+        * If we need to support user-supplied session objects, those should be passed in to the constructor
+    * It may or may not be a good idea for the client to be usable as a context manager that closes its session on exit.
+        * Con: Closing the session would affect `RemoteDandiset` etc. instances for the client (see below), making them unusable.
+    * Rename the `send_request()` method to `request()` to parallel `requests.Session.request()`
+    * Rename the `parameters` argument to the HTTP methods to `params` to match `requests`
+    * The `api_url` argument to `DandiAPIClient`'s constructor should default to the URL for the official Dandiarchive instance (which should also be available as a constant)
+    * Give `DandiAPIClient` a classmethod `for_dandi_instance(cls, instance: Union[str, dandi_instance], token=None, authenticate=True)` (Rethink name) that takes either the name of a Dandi instance or an actual `dandi_instance`, fetches a token from the environment, keychain, and/or user if one is not given and `authenticate` is true, and returns a `DandiAPIClient` instance
+    * Most interaction with API resources should be via new `RemoteDandiset`, `Version`, and `RemoteAsset` classes and their methods
+        * These classes (except `Version`?) will all contain a private `_client` attribute referring back to the associated `DandiAPIClient` instance.
+    * Changes to `DandiAPIClient` methods:
+        * Add a `get_dandiset(dandiset_id: Union[str, RemoteDandiset], version_id=None) -> RemoteDandiset` method
+            * **To discuss:** Should the default `version` be "draft" or the most recent version?  Either way, there should be another method that explicitly gets the other option.
+            * **To discuss:** Should this take a `lazy=False` parameter that, if true, constructs a `RemoteDandiset` containing only an identifier, without making any requests at the time of creation?  This would allow calling a `RemoteDandiset`'s methods without having to make a request for the Dandiset's data beforehand.  If & when any non-identifiers attributes of a lazy `RemoteDandiset` are requested, a request is made at that point and the results cached.
+        * Add a `get_all_dandisets() -> Iterator[RemoteDandiset]` method
+            * **To discuss:** What should these objects' `Version`s be?
+        * Retype `create_dandiset` to `create_dandiset(name: str, metadata: DandiMeta) -> RemoteDandiset`
```

Comment:
I think we should remove explicit `name` from the signature.  Should be present in metadata anyways.
It does make it less user-friendly since hard to figure out from the function signature though :-/

A possible choice could be to make it `create_dandiset(*metadata)` and mint the docstring listing the fields based on `DandiMeta` from which we know what is mandatory and what is optional.  I do not know though if it is possible to autogenerate the signature though for extra user/IDE's convenience.
But that would remove (or complicate) possibility to provide extra options which aren't metadata, and not sure how python typing would take it.

So may be just making it `create_dandiset(metadata: DandiMeta)` and minting docstring with a nice description of fields in `metadata` would be the reasonable compromise.

Reply:
`name` removed.